### PR TITLE
ARROW-4255: [C++] Eagerly initialize name_to_index_ to avoid race

### PR DESCRIPTION
**Pull request by gnpdt**

### CODE REVIEW

**Review by bkietz — APPROVED** [review]

File: cpp/src/arrow/type.cc [hunk 1]
```diff
@@ -219,6 +219,15 @@ std::string UnionType::ToString() const {
 // ----------------------------------------------------------------------
 // Struct type
 
+StructType::StructType(const std::vector<std::shared_ptr<Field>>& fields)
+  : NestedType(Type::STRUCT)
+{
+  children_ = fields;
```
Comment (emkornfield) [inline cpp/src/arrow/type.cc]: not a lot of code but would it make sense to refactor the population into a utility method?

**Review by pitrou — APPROVED** [review]
+1. Wil merge after CI passes.

**Comment by xhochy** [comment]
@gnpdt Do you have an account on https://issues.apache.org/jira/projects/ARROW so I can assign the issue to you so that you get credits in the Changelog?

**Comment by gnpdt** [comment]
Just created one (same username). Thanks!

**Pull request closed without merging**